flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicKeySelector.java

```diff
@@ -0,0 +1,384 @@
+              WRITE_DISTRIBUTION_MODE);
+          return equalityFieldKeySelector(
+              tableName, schema, equalityFields, writeParallelism, maxWriteParallelism);
+        }
+
+      default:
+        throw new IllegalArgumentException("Unrecognized " + WRITE_DISTRIBUTION_MODE + ": " + mode);
+    }
+  }
+
+  private static NonThrowingKeySelector<RowData, Integer> equalityFieldKeySelector(
+      String tableName,
+      Schema schema,
+      List<String> equalityFields,
+      int writeParallelism,
+      int maxWriteParallelism) {
+    return new TargetLimitedKeySelector(
+        new EqualityFieldKeySelector(
+            schema,
+            FlinkSchemaUtil.convert(schema),
+            DynamicRecordProcessor.getEqualityFieldIds(equalityFields, schema)),
+        tableName.hashCode(),
+        writeParallelism,
+        maxWriteParallelism);
+  }
+
+  private static NonThrowingKeySelector<RowData, Integer> partitionKeySelector(
+      String tableName,
+      Schema schema,
+      PartitionSpec spec,
+      int writeParallelism,
+      int maxWriteParallelism) {
+    NonThrowingKeySelector<RowData, String> inner =
+        new PartitionKeySelector(spec, schema, FlinkSchemaUtil.convert(schema));
+    return new TargetLimitedKeySelector(
+        in -> inner.getKey(in).hashCode(),
+        tableName.hashCode(),
+        writeParallelism,
+        maxWriteParallelism);
+  }
+
+  private static NonThrowingKeySelector<RowData, Integer> tableKeySelector(
+      String tableName, int writeParallelism, int maxWriteParallelism) {
+    return new TargetLimitedKeySelector(
+        new RoundRobinKeySelector<>(writeParallelism),
+        tableName.hashCode(),
+        writeParallelism,
+        maxWriteParallelism);
+  }
+
+  /**
+   * Generates a new key using the salt as a base, and reduces the target key range of the {@link
+   * #wrapped} {@link NonThrowingKeySelector} to {@link #writeParallelism}.
+   */
+  private static class TargetLimitedKeySelector
+      implements NonThrowingKeySelector<RowData, Integer> {
+    private final NonThrowingKeySelector<RowData, Integer> wrapped;
+    private final int writeParallelism;
+    private final int[] distinctKeys;
+
+    @SuppressWarnings("checkstyle:ParameterAssignment")
+    TargetLimitedKeySelector(
+        NonThrowingKeySelector<RowData, Integer> wrapped,
+        int salt,
+        int writeParallelism,
+        int maxWriteParallelism) {
+      if (writeParallelism > maxWriteParallelism) {
+        LOG.warn(
+            "writeParallelism {} is greater than maxWriteParallelism {}. Capping writeParallelism at {}",
+            writeParallelism,
+            maxWriteParallelism,
+            maxWriteParallelism);
+        writeParallelism = maxWriteParallelism;
+      }
+      this.wrapped = wrapped;
+      this.writeParallelism = writeParallelism;
+      this.distinctKeys = new int[writeParallelism];
+
+      // Ensures that the generated keys are always result in unique slotId
```

Comment:
can you explain the usage of `KeyGroupRangeAssignment` to probe the slots? what does the slotId mean here? why do we use `KeyGroupRangeAssignment`?

Reply:
SlotId is the same as the subtasks id, but zero-indexed (0..parallelism-1). We probe for the key group which is managed by the subtask that we target. Round-robin is the default behavior for DistributionMode NONE. In order to support setting the writeParallelism for each table independently, we have to keyBy even for NONE distribution mode, to ensure that each table gets its dedicated set of subtasks.

I'm also described this process in https://github.com/apache/iceberg/pull/12424#discussion_r2002945179.